airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java

```diff
@@ -0,0 +1,100 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.commons.json;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Preconditions;
+
+public class JsonSecretsProcessor {
+
+  public static String AIRBYTE_SECRET_FIELD = "airbyte_secret";
+  private static String PROPERTIES_FIELD = "properties";
+
+  /**
+   * Returns a copy of the input object wherein any fields annotated with "airbyte_secret" in the
+   * input schema are removed.
+   * <p>
+   * TODO this method only removes secrets at the top level of the configuration object. It does not support the keywords anyOf, allOf, oneOf, not, and
+   * dependencies. This will be fixed in the future.
+   *
+   * @param schema Schema containing secret annotations
+   * @param obj    Object containing potentially secret fields
+   * @return
+   */
+  public JsonNode removeSecrets(JsonNode obj, JsonNode schema) {
+    assertValidSchema(schema);
+    Preconditions.checkArgument(schema.isObject());
+
+    ObjectNode properties = (ObjectNode) schema.get(PROPERTIES_FIELD);
+    JsonNode copy = obj.deepCopy();
+    for (String key : Jsons.keys(properties)) {
+      if (isSecret(properties.get(key))) {
+        ((ObjectNode) copy).remove(key);
+      }
+    }
+
+    return copy;
+  }
+
+  /**
+   * Returns a copy of the destination object in which any secret fields (as denoted by the input
+   * schema) found in the source object are added.
+   * <p>
+   * TODO this method only absorbs secrets at the top level of the configuration object. It does not support the keywords anyOf, allOf, oneOf, not, and
+   * dependencies. This will be fixed in the future.
+   *
+   * @param src    The object potentially containing secrets
+   * @param dst    The object to absorb secrets into
+   * @param schema
+   * @return
+   */
+  public JsonNode copySecrets(JsonNode src, JsonNode dst, JsonNode schema) {
+    assertValidSchema(schema);
+    Preconditions.checkArgument(dst.isObject());
+    Preconditions.checkArgument(src.isObject());
+
+    ObjectNode dstCopy = dst.deepCopy();
+
+    ObjectNode properties = (ObjectNode) schema.get(PROPERTIES_FIELD);
+    for (String key : Jsons.keys(properties)) {
+      if (isSecret(properties.get(key)) && src.has(key)) {
+        dstCopy.set(key, src.get(key));
+      }
+    }
+
+    return dstCopy;
+  }
+
+  private static boolean isSecret(JsonNode obj) {
+    return obj.isObject() && obj.has(AIRBYTE_SECRET_FIELD) && obj.get(AIRBYTE_SECRET_FIELD).asBoolean();
+  }
+
+  private static void assertValidSchema(JsonNode node) {
```

Comment:
why do we want to throw is a schema doesn't match our expectations? don't we just want to do nothing? e.g. if there is no top level properties and it is just a oneOf we don't want to fail do we?

Reply:
fair enough. will fix